kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/usagelimits/ratelimits/DefaultRateLimiter.kt

```diff
@@ -0,0 +1,197 @@
+     */
+    public override suspend fun checkCommandRatelimit(context: DiscriminatingContext): Boolean {
+        val currentTimeMillis = System.currentTimeMillis()
+        val numerOfTypes = CachedUsageLimitType.values().size
+        val hitRateLimits = ArrayList<Triple<RateLimitType, UsageHistory, RateLimit>>(numerOfTypes)
+        var shouldSendMessage = true
+
+        for (type in CachedUsageLimitType.values()) {
+            val rateLimit = rateLimitProvider.getRateLimit(context, type)
+            if (!rateLimit.enabled) continue
+
+            val usageHistory = type.getUsageHistory(context)
+            val encapsulateStart = currentTimeMillis - rateLimit.duration.inWholeMilliseconds
+
+            // keeps only the usageHistory which is in the rateLimit's range,
+            // usage that doesn't affect the ratelimit is thus discarded.
+            usageHistory.removeExpiredUsages(encapsulateStart)
+
+            // keeps only the crossedLimits which are in the rateLimit's range.
+            usageHistory.removeExpiredCrossedLimits(encapsulateStart)
+
+            if (usageHistory.usages.size > rateLimit.limit) {
+                hitRateLimits.add(Triple(type, usageHistory, rateLimit))
+                if (!shouldSendMessage(usageHistory, rateLimit, type)) shouldSendMessage = false
+                usageHistory.addCrossedLimit(currentTimeMillis)
+            } else {
+                usageHistory.addUsage(currentTimeMillis)
+            }
+
+            type.setUsageHistory(context, usageHistory)
+        }
+
+        if (shouldSendMessage) {
+            val (maxType, maxUsageHistory, maxRateLimit) = hitRateLimits.maxByOrNull { (_, usageHistory, rateLimit) ->
+                rateLimit.duration.minus(usageHistory.usages.first().milliseconds)
+            } ?: return false
+            sendRateLimitedMessage(context, maxType, maxUsageHistory, maxRateLimit)
+        }
+
+        return hitRateLimits.isNotEmpty()
+    }
+
+    /**
+     * @return true if a ratelimit message should be sent, false otherwise.
+     */
+    public override suspend fun shouldSendMessage(
+        usageHistory: UsageHistory,
+        rateLimit: RateLimit,
+        type: RateLimitType,
+    ): Boolean =
+        System.currentTimeMillis() - (usageHistory.crossedLimits.lastOrNull() ?: 0) > backOffTime.inWholeMilliseconds
+
+    /** Returns a message with info about what ratelimit has been hit. **/
+    public open suspend fun getMessage(
+        context: DiscriminatingContext,
+        discordTimeStamp: String,
+        type: RateLimitType,
+    ): String {
+        val locale = context.locale()
+        val translationsProvider = context.event.command.translationsProvider
+        val commandName = context.event.command.getFullName(locale)
+
+        /** "You are being RateLimited until $discordTimeStamp, please stop spamming.\n" +
+        " ${rateLimit.limit} actions per ${rateLimit.duration.toIsoString()} is the limit." **/
```

Comment:
What's this comment for? I assume it's been moved to a translation string already, right?